Track completion status of timeouts and allow retrieval of scheduled and completed timeouts

shawmanz32na opened the pull request:
This adds two new methods to the `timeoutCollection` object: `getScheduled` and `getCompleted`. The `Timeout` object has been expanded to have a `status` field, and has been given a constructor to quickly/easily set the `timestamp` and `status`. 

I'm not fully confident that my `_getWrappedHandler` method is the cleanest implementation of marking the `Timeout` complete when triggered, so I welcome your comments, suggestions, or a different implementation. Additionally, I suspect that it doesn't respect any additional parameters passed into timeout, but don't have a valid use case with which to test. Perhaps you can double-check me here, and maybe offer a better solution?

Refs #3

shawmanz32na committed:
Track completion status of timeouts and allow retrieval of scheduled and completed timeouts

Refs #3

bargoldi commented on src/timeout/timeout-collection.model.ts:
```diff
@@ -1,16 +1,23 @@
 		return id;
 	}
 
+	private _getWrappedHandler(handler: Function): Function {
```
Move the private method to the end please

bargoldi commented on src/timeout/timeout-collection.model.ts:
```diff
@@ -1,16 +1,23 @@
+	private _getWrappedHandler(handler: Function): Function {
+		return (() => {
+			this._timeoutCollection[this._getTimeoutIndexByHandler(handler)].status = TimeoutStatus.Completed;
+			return handler.apply(window);
```
add a separating line before the return statement please

shawmanz32na replied:
I'm afraid I don't understand... If you give me a snippet of the desired code, I'll drop it in.

bargoldi replied:
Code

return blabla

I mean add an empty line between 16 and 17

shawmanz32na replied:
Okie dokie. I don't see that convention used elsewhere, but you're the boss!

bargoldi commented on src/timeout/timeout-status.model.ts:
```diff
@@ -0,0 +1,4 @@
+export enum TimeoutStatus {
+  Scheduled,
+  Completed,
```
Remove trailing comma please

bargoldi commented on src/timeout/timeout-collection.model.ts:
```diff
@@ -25,6 +32,18 @@ export class TimeoutCollection {
+
+	public getCompleted(): Timeout[] {
+		return this._timeoutCollection.filter((value: Timeout) => {
+			return value.status == TimeoutStatus.Completed;
```
fix to === please

shawmanz32na committed:
Style improvements per code review in #4

Refs #3, #4

zivpolack commented on src/timeout/timeout-collection.model.ts:
```diff
@@ -60,6 +53,14 @@ export class TimeoutCollection {
 		this._timeoutCollection = [];
 	}
 
+	private _getWrappedHandler(handler: Function): Function {
```
This can potentially cause problems when calling mutliple timeouts with the same handler.. (When the bigger indexed timeout is executed prior to the smaller indexed one)

shawmanz32na replied:
Good catch.

Unfortunately, I struggled all morning trying to figure out how to wire the "completing" of the `Timeout` objects into the delegation to the default `setTimeout`, and still don't really have any good ideas for how to mitigate this issue... the `id` is the unique field, but I don't have it until after I schedule a timeout, at which point I've already committed to a handler...

Do you have any ideas for how to better lookup the `Timeout` object to remove when the timeout completes?

zivpolack replied:
My initial thought on this was to have a `_scheduledTimeoutCollection` in addition to the normal one, and once `timeoutCollection.add()` is called, it would push the `new Timeout` object to both collections, and in addition create an `originalSetTimeout` (same as window's one) to pop that timeout from `_scheduledTimeoutCollection` after the same time delay.

I do like your solution better though, if we are able to find a reasonable way to distinguish the timeouts without having the `id`

shawmanz32na replied:
I've had a brain blast of sorts...

We _could_ maintain a time-events-manager-unique UUID in each `Timeout`. We could generate the UUID as part of the Timeout constructor, and use that UUID to reference the `Timeout` from the "completion" logic. Then, we could set the normal `id` on the `Timeout` whenever we get it back from the default `setTimeout`, and use that id when we delegate to the default `clearTimeout` method.

What do you think?

If you're ameniable to the idea, how do you feel about requiring other dependencies, such as https://github.com/kelektiv/node-uuid ? I _could_ add a psuedo-UUID generator, but I'd rather rely on one that's pretty robust to get truly unique UUIDs...

zivpolack replied:
I tried avoiding that at first since the original objects already have a UUID, but seems like it's the only reasonable way if we want to implement the status solution.

@bargoldi anything to add?

shawmanz32na committed:
Fix for the possibility of marking the wrong Timeout complete

There are complications using the default `setTimeout`-provided id when
referencing the Timeout to mark complete, since we don't get the id from
the default `setTimeout` until after we've committed to a handler. To
work around the issue, an app-specific random UUID is set for each
Timeout object, which is used in the handler to look up which Timeout to
mark as complete.

Refs #3, #4
